pkg/controller/deployment/deployment_controller.go

```diff
@@ -543,7 +543,18 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 		return dc.syncStatusOnly(d)
 	}
 
-	err = dc.classifyReplicaSets(deployment)
+	// Run cleanup policy before adopting if there is no rollback request that needs to happen. Doing so
+	// decouples the cleanup policy from the strategies where it gets executed and may block because the
+	// strategy can't proceed.
```

Comment:
The deployment will adopt replica sets in `classifyReplicaSets`. The cleanup policy gets executed after deployments have successfully completed a rollout but when a rollout blocks (for example, the new replica set cannot be created), the cleanup policy is not executed. I will update the comment to make it clearer.